File: zig/enc-legacy/Stats.zig
const std = @import("std");
const enc = @import("./enc.zig");

const Self = @This();

pub const Stat = enum {
    min,
    max,
    is_constant,
    is_sorted,
    avg_run_length, // TODO(ngates): change to run_count
    bit_width_freq,
    true_count,
};

const StatsMap = std.EnumMap(enc.Stats.Stat, enc.Scalar);

values: StatsMap = StatsMap{},

pub fn deinit(self: *Self) void {
    var iter = self.values.iterator();
    while (iter.next()) |entry| {
        entry.value.deinit();
    }
}

pub fn get(self: Self, stat: Stat) ?enc.Scalar {
    return self.values.get(stat);
}

pub fn put(self: *Self, stat: Stat, scalar: enc.Scalar) void {
    self.values.put(stat, scalar);
}

File: vortex-array2/src/stats.rs
use std::collections::HashMap;

use vortex::scalar::Scalar;
use vortex_error::VortexResult;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum Stat {
    BitWidthFreq,
    TrailingZeroFreq,
    IsConstant,
    IsSorted,
    IsStrictSorted,
    Max,
    Min,
    RunCount,
    TrueCount,
    NullCount,
}

pub trait ArrayStatistics {
    fn statistics(&self) -> &(dyn Statistics + '_) {
        &EmptyStatistics
    }
}

pub trait ArrayStatisticsCompute {
    /// Compute the requested statistic. Can return additional stats.
    fn compute_statistics(&self, _stat: Stat) -> VortexResult<HashMap<Stat, Scalar>> {
        Ok(HashMap::new())
    }
}

pub trait Statistics {
    fn compute(&self, stat: Stat) -> Option<Scalar>;
    fn get(&self, stat: Stat) -> Option<Scalar>;
    fn set(&self, stat: Stat, value: Scalar);
    fn to_map(&self) -> HashMap<Stat, Scalar>;
}

impl dyn Statistics + '_ {
    pub fn compute_as<T: TryFrom<Scalar>>(&self, stat: Stat) -> Option<T> {
        self.compute(stat).and_then(|s| T::try_from(s).ok())
    }

    pub fn get_as<T: TryFrom<Scalar>>(&self, stat: Stat) -> Option<T> {
        self.get(stat).and_then(|s| T::try_from(s).ok())
    }
}

pub struct EmptyStatistics;
impl Statistics for EmptyStatistics {
    fn compute(&self, _stat: Stat) -> Option<Scalar> {
        None
    }
    fn get(&self, _stat: Stat) -> Option<Scalar> {
        None
    }
    fn set(&self, _stat: Stat, _value: Scalar) {}
    fn to_map(&self) -> HashMap<Stat, Scalar> {
        HashMap::default()
    }
}
